c2rust-analyze/src/main.rs

```diff
@@ -321,6 +321,46 @@ where
     }
 }
 
+fn mark_foreign_fixed<'tcx>(
+    gacx: &mut GlobalAnalysisCtxt<'tcx>,
+    gasn: &mut GlobalAssignment,
+    tcx: TyCtxt<'tcx>,
+) {
+    // FIX the inputs and outputs of function declarations in extern blocks
+    for did in gacx.foreign_mentioned_tys.clone() {
+        if let DefKind::Fn = tcx.def_kind(did) {
+            let sig = tcx.erase_late_bound_regions(tcx.fn_sig(did));
+            let inputs = sig
+                .inputs()
+                .iter()
+                .map(|&ty| gacx.assign_pointer_ids_with_info(ty, PointerInfo::ANNOTATED))
+                .collect::<Vec<_>>();
+            for input in inputs {
+                make_ty_fixed(gasn, input);
+            }
+
+            let output = gacx.assign_pointer_ids_with_info(sig.output(), PointerInfo::ANNOTATED);
+            make_ty_fixed(gasn, output)
+        }
+    }
```

Comment:
```suggestion
    let foreign_mentioned_tys = mem::take(&mut gacx.foreign_mentioned_tys);
    for ty in foreign_mentioned_tys
        .iter()
        .filter(|&did| tcx.def_kind(did) == DefKind::Fn)
        .map(|did| tcx.fn_sig(did))
        .map(|sig| tcx.erase_late_bound_regions(sig))
        .flat_map(|sig| sig.inputs_and_output)
    {
        let lty = gacx.assign_pointer_ids_with_info(ty, PointerInfo::ANNOTATED);
        make_ty_fixed(gasn, lty);
    }
    let _ = mem::replace(&mut gacx.foreign_mentioned_tys, foreign_mentioned_tys);
```

The `mem::take`/`mem::replace` (needs a `use std::mem`) would avoid the need to `.clone()` the `HashSet` while still getting around the borrow checker without refactoring things.  And we shouldn't need to `.collect()` the `inputs` either, I think.

Reply:
No longer necessary with 7f8bddbe